applications/StructuralMechanicsApplication/custom_response_functions/adjoint_conditions/adjoint_semi_analytic_base_condition.cpp

```diff
@@ -0,0 +1,410 @@
+
+        }
+        else {
+            KRATOS_ERROR << "Unsupported output variable." << std::endl;
+        }
+
+        KRATOS_CATCH("")
+    }
+
+    template <class TPrimalCondition>
+    void AdjointSemiAnalyticBaseCondition<TPrimalCondition>::CalculateSensitivityMatrix(const Variable<double>& rDesignVariable,
+                                            Matrix& rOutput,
+                                            const ProcessInfo& rCurrentProcessInfo)
+    {
+        KRATOS_TRY
+
+        ProcessInfo process_info = rCurrentProcessInfo;
+        const SizeType number_of_nodes = this->GetGeometry().size();
+        const SizeType dimension =  this->GetGeometry().WorkingSpaceDimension();
+        const SizeType local_size = number_of_nodes * dimension;
+
+        // this is to derive w.r.t. to variables like PRESSURE
+        if(this->Has(rDesignVariable)) {
+            if ((rOutput.size1() != 1) || (rOutput.size2() != local_size)) {
+                rOutput.resize(1, local_size, false);
+            }
+            noalias(rOutput) = ZeroMatrix(1,local_size);
+
+            const double delta = this->GetPerturbationSize(rDesignVariable, rCurrentProcessInfo);
+            Vector RHS;
+            Vector perturbed_RHS;
+            this->CalculateRightHandSide(RHS, process_info);
+
+            const auto design_variable_value = this->pGetPrimalCondition()->GetValue(rDesignVariable);
+
+            // perturb design variable
+            this->pGetPrimalCondition()->SetValue(rDesignVariable, (design_variable_value + delta));
+            this->pGetPrimalCondition()->CalculateRightHandSide(perturbed_RHS, process_info);
+
+            row(rOutput, 0) = (perturbed_RHS - RHS) / delta;
+
+            // unperturb design variable
+            this->pGetPrimalCondition()->SetValue(rDesignVariable, design_variable_value);
+        }
+        else {
+            if ((rOutput.size1() != 0) || (rOutput.size2() != local_size)) {
+                rOutput.resize(0, local_size, false);
+            }
+            noalias(rOutput) = ZeroMatrix(0, local_size);
+        }
+
+        KRATOS_CATCH( "" )
+    }
+
+    template <class TPrimalCondition>
+    void AdjointSemiAnalyticBaseCondition<TPrimalCondition>::CalculateSensitivityMatrix(const Variable<array_1d<double,3> >& rDesignVariable,
+                                            Matrix& rOutput,
+                                            const ProcessInfo& rCurrentProcessInfo)
+    {
+        KRATOS_TRY
+
+        ProcessInfo process_info = rCurrentProcessInfo;
+        const SizeType number_of_nodes = this->GetGeometry().size();
+        const SizeType dimension = this->GetGeometry().WorkingSpaceDimension();
+        const SizeType local_size = number_of_nodes * dimension;
+        Vector RHS;
+        Vector perturbed_RHS = Vector(0);
+        const double delta = this->GetPerturbationSize(rDesignVariable, rCurrentProcessInfo);
+
+        this->CalculateRightHandSide(RHS, process_info);
```

Comment:
It should be possible to use the ElementFiniteDifferenceUtility here as well.

One would need to template the functions there (and maybe remove the "Element" from the name)

Reply:
You are right. For SHAPE_SENSITIVITY it should be possible. But not for the other design variables, since the elemental variables are part of the properties and the conditional variables are typically part of the non-historical database.

Reply:
I see. As you prefer.

Reply:
But it makes sense to open the element finite difference utility for conditions. I started now to template at least for SHAPE_SENSITIVITY. I will also rename to `FiniteDifferenceUtility`.